pint_server/app.py

```diff
@@ -212,7 +312,42 @@ def get_provider_servers_types(provider):
         return [{'name': 'smt'}, {'name': 'regionserver'}]
 
 
-def get_provider_regions(provider):
+# generate a provider specific hash that is dependent on the data
+# versions of the respective provider's category tables, if they
+# exist.
+# Optionally specify a category of table as a string or a list of
+# categories if desired. Defaults to matching images and servers
+# tables for a provider.
+def get_provider_hash(provider, categories=None):
+    if categories is None:
+        categories = ['images', 'servers']
+    elif type(categories) is str:
+        categories = [categories]
+
+    provider_tables = [provider + c
+                       for c in categories]
+
+    found_versions = (v.version
+                for v in VersionsModel.query.filter(
+                    VersionsModel.tablename.in_(provider_tables))
+                    .order_by(VersionsModel.tablename)
+    )
+
+    # include the provider name in the hashed content
+    return hash((provider, found_versions))
+
+
+# caches results of retrieving regions list for providers; cache_hash
+# argument can be used to ensure stale entries get expired. Maintains
+# at most one active cache entry for each provider. This caching works
+# for both short lived instances handling a single request, as it will
+# cache the region list for a specific provider so it only needs to be
+# calulated once, as well as for persistent server instances, even if
+# the DB data changes.
+@lru_cache(maxsize=len(PROVIDER_IMAGES_MODEL_MAP))
```

Comment:
I think in-memory caching will be marginally beneficial as the lambda container have a limited lifespan (of 15 minutes?) https://aws.amazon.com/lambda/faqs/.

Reply:
So the in-memory caching means we only really do the provider regions call once for a given region, with the same data versioning hash. It has a small performance gain for an active stream of requests, and allows for moving the region validity checking to the request entry point handler, without the negative impact of having to do the actual query twice if the region specified was valid.

Reply:
The question is also how we configure the lambda function, we could end up with a cache for each provider in memory, is the lambda function configured to handle this?

Reply:
My understanding is that the lambda function instance sticks around for a few mins before being disposed of, so if subsequent queries arrive during that timeframe then then could potentially benefit from the cached values. If there is concern about the size of the cached values (which isn't necessarily too large right now) we could reduce the cache size down to 1 so that the caching benefit is primarily focused the current request, and only benefits subsequent requests if they are for the same provider.